Kernel/Theory.hpp

```diff
@@ -54,77 +80,150 @@ class DivByZeroException         : public ArithmeticException
   DivByZeroException() : ArithmeticException("divided by zero"){} 
 };
 
+enum class Sign : uint8_t {
+  Zero = 0,
+  Pos = 1,
+  Neg = 2,
+};
+
+inline std::ostream& operator<<(std::ostream& out, Sign const& self)
+{
+  switch(self) {
+    case Sign::Zero: return out << "0";
+    case Sign::Pos: return out << "+";
+    case Sign::Neg: return out << "-";
+  }
+  ASSERTION_VIOLATION
+}
+
+template<class T> struct MpzToMachineInt;
+
+template<> struct MpzToMachineInt<long> {
+  using Int = long;
+  static bool fits(mpz_t const self) { return mpz_fits_sint_p(self);  }
+  static Int cvt(mpz_t const self) { return mpz_get_si(self);  }
+};
+#define CONV_FROM_NUM_LIMITS(Long) \
+  static Option<Int> _cvt(mpz_t const self) {  \
+    if (!MpzToMachineInt<Long>::fits(self)) \
+      return {}; \
+    auto v = MpzToMachineInt<Long>::cvt(self); \
+    auto lim = std::numeric_limits<Int>{}; \
+    if (lim.min() <= v && v <= lim.max()) { \
+      return some(Int(v)); \
+    } else { \
+      return {}; \
+    } \
+  } \
+  static bool fits(mpz_t const self) { return _cvt(self).isSome(); } \
+  static Int cvt(mpz_t const self) { return _cvt(self).unwrap(); } \
+
+template<> struct MpzToMachineInt<unsigned long> {
+  using Int = unsigned long;
+
+  static Int truncate(mpz_t const self) { return mpz_get_ui(self); }
+  static bool fits(mpz_t const self) { return mpz_sgn(self) >= 0 && mpz_fits_uint_p(self);  }
+  static Int cvt(mpz_t const self) { return mpz_get_ui(self);  }
+};
+
+template<> struct MpzToMachineInt<int> {
+  using Int = int;
+  CONV_FROM_NUM_LIMITS(long)
+};
+
+template<> struct MpzToMachineInt<unsigned> {
+  using Int = unsigned;
+  CONV_FROM_NUM_LIMITS(unsigned long)
+  static Int truncate(mpz_t const self) { return mpz_get_ui(self); }
+};
+
+
+
 class IntegerConstantType
 {
+  mpz_t _val;
 public:
   static TermList getSort() { return AtomicSort::intSort(); }
 
-  typedef int InnerType;
-
-  IntegerConstantType() {}
-  IntegerConstantType(IntegerConstantType&&) = default;
-  IntegerConstantType(const IntegerConstantType&) = default;
-  IntegerConstantType& operator=(const IntegerConstantType&) = default;
-  constexpr IntegerConstantType(InnerType v) : _val(v) {}
+  IntegerConstantType() { mpz_init(_val); }
+  explicit IntegerConstantType(int v) : IntegerConstantType() { mpz_set_si(_val, v); }
   explicit IntegerConstantType(const std::string& str);
 
+  IntegerConstantType(IntegerConstantType     && o) : IntegerConstantType() { mpz_swap(_val, o._val); }
+  IntegerConstantType(IntegerConstantType const& o) : IntegerConstantType() {  mpz_set(_val, o._val); }
```

Comment:
Does this make a deep copy of `_val`? Otherwise when one destructor runs both will be invalidated.

Reply:
Do you mean `mpz_set`? That one makes a deep copy.